src/mahalanobis.jl

```diff
@@ -1,19 +1,19 @@
 # Mahalanobis distances
 
-struct Mahalanobis{T} <: Metric
-    qmat::AbstractMatrix{T}
+struct Mahalanobis{T,M<:AbstractMatrix{T}} <: Metric
+    qmat::M
 end
 
-struct SqMahalanobis{T} <: SemiMetric
-    qmat::AbstractMatrix{T}
+struct SqMahalanobis{T,M<:AbstractMatrix{T}} <: SemiMetric
+    qmat::M
 end
 
-result_type(::Mahalanobis{T}, ::Type, ::Type) where {T} = T
-result_type(::SqMahalanobis{T}, ::Type, ::Type) where {T} = T
+result_type(::Mahalanobis{T,M}, ::Type, ::Type) where {T,M} = T
+result_type(::SqMahalanobis{T,M}, ::Type, ::Type) where {T,M} = T
```

Comment:
No need to mention `M` here. Trailing types can be left implicit.

Reply:
Should this be modified before the PR can be merged @dkarrasch or it can be left that way? I personally find the explicit version with the 2 type parameters better.

Reply:
Sure, this is just a style comment. OTOH, it's Julia style not to spell out all parameters. There is one other thing that confuses me. Is it really intentional to have the eltype of the matrix determine the `result_type`? What if you have an integer matrix, but float vectors? Shouldn't the result_type be sufficiently large to hold `one(Ta)*one(T)*one(Tb)[ + one(Ta)*one(T)*one(Tb)]`?

Reply:
That is a good point. I guess #179 also tries to fix this issue, right? The result type should be a combination of all argument types with `promote_type`.

Reply:
@johnnychen94 since you are up-to-date with the linked issue above, what is your suggestion for the `result_type` of Mahalanobis here? Can it just be the following?

```julia
result_type(::Mahalanobis{T,M}, ::AbstractVector{T1}, ::AbstractVector{T2}) where {T,T1,T2} = reduce(promote_type, (T,T1,T2))
```

Reply:
No need to wrap it with `reduce` here:

```julia
promote_type(T, T1, T2)
```

would be fine.

> What if you have an integer matrix, but float vectors?

A good example, probably we can also strip off the `T` type in the struct definition, i.e.,

```julia
struct Mahalanobis{M<:AbstractMatrix} <: Metric
    qmat::M
end
```

Both of these should be seen as breaking changes (e.g., memory allocation is sensitive to result type), so I think it's safer to get it in v0.10 releases.

Reply:
@rmcaixeta could you please implement the proposed changes? Update the Mahalanobis struct as suggested by @johnnychen94 to contain a single type parameter `M` and update the result_type methods to something like

```Julia
result_type(d::Mahalanobis, ::AbstractVector{T1}, ::AbstractVector{T2}) = promote_type(eltype(d.qmat), T1, T2)
```

Reply:
I fixed a mild typo above. Also, tests will need to be adjusted, because currently they are testing if the result type is equal to the eltype of `Q`. I hope the `F64` thing will go through smoothly.

Reply:
Thanks. PR updated

Reply:
Wait, I overlooked that. I think it should be
```julia
result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2} = promote_type(eltype(d.qmat), T1, T2)
```
to stay in line with the previous signature.

Reply:
@dkarrasch the last time I checked the internals of Distances.jl the `result_type` function took the distance and two vectors as input. Did that change? The PR was updated correctly by @rmcaixeta assuming that these are still the inputs.

Reply:
No. Look at the generic fallback:

https://github.com/JuliaStats/Distances.jl/blob/777f02307d152adc8dedca27df736a61efd975b8/src/generic.jl#L36

For specific metrics, we need to specialize `result_type` for types, not for arrays, not to mention for vectors. This is also the current signature

https://github.com/JuliaStats/Distances.jl/blob/777f02307d152adc8dedca27df736a61efd975b8/src/mahalanobis.jl#L11-L12

which simply ignores the eltypes of the arrays, and takes the eltype of `Q`.

Reply:
Done

Reply:
Hmmm, (sorry for being so picky, but I want to get it really right): what if the elements of the vectors are matrices (or colorants or whatever), and (potentially) `mahalanobis` is implemented via 3-arg `dot`? Is the `promote_type` approach gonna work? @johnnychen94 @devmotion 

We use
```julia
function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
    (axes(x)..., axes(y)...) == axes(A) || throw(DimensionMismatch())
    T = typeof(dot(first(x), first(A), first(y)))
...
```
to determine the result type in 3-arg `dot`. We could go back to the array-based result type determination, like
```julia
result_type(d::Mahalanobis, x::AbstractArray{T1}, y::AbstractArray{T2}) where {T1,T2} = typeof((first(x)-first(y))*first(d.qmat))
```
? I apologize once more for my back and forth, but since we're dropping the strict eltype requirement, we need to have it work for a wild variety of eltypes.

Reply:
IMO it is better to not use `promote_type` but rather a similar approach as in other places currently by defining
```julia
result_type(d::SqMahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2} = typeof(oneunit(eltype(d.qmat)) * oneunit(T1) * oneunit(T2))
```
In contrast to `promote_type`, this will work with unitful numbers as well.

I also don't think one should worry about `a` and `b` being vectors of matrices etc. but rather adopt the convention that the type based definitions of `result_type` *always* operate with the element types. Supporting vectors of matrices, colorants, etc. should then be possible by just defining
```julia
result_type(d::Metric, a::MyCustomType, b::MyCustomType) = result_type(d, get_the_eltype(a), get_the_eltype(b))
```
similar to the existing definitions for `AbstractArray`.

Reply:
Sounds good. I changed a detail, `oneunit(T1) - oneunit(T2)`, because this is how it's computed, and this may yield a new type (I'm thinking of `T1 == T2 == Bool`).

Reply:
Ah, thanks for noticing! Actually, to correct my statement about unitful numbers, it should be
```julia
function result_type(d::SqMahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = oneunit(T1) - oneunit(T2)
    return typeof(z * oneunit(eltype(d.qmat)) * z)
end
```
since for unitful numbers `typeof(z * z) != typeof(z)`.

Reply:
When I introduced this approach in #140, my main focus was that method dispatching becomes way too confusing to write correctly, and it was very easy to introduce method ambiguity. For example, https://github.com/JuliaImages/ImageDistances.jl/issues/28#issuecomment-495010897

Given that Distances.jl computes numerical data, container type (array or colorant) is not needed to infer the result type in cases so there is no need to dispatch on arrays.

Reply:
Nice. @dkarrasch , PR updated with the 2nd correction:

```julia
function result_type(d::Mahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = oneunit(T1) - oneunit(T2)
    return typeof(z * oneunit(eltype(d.qmat)) * z)
end

function result_type(d::SqMahalanobis, ::Type{T1}, ::Type{T2}) where {T1,T2}
    z = oneunit(T1) - oneunit(T2)
    return typeof(z * oneunit(eltype(d.qmat)) * z)
end
```